File: creational/singleton/kotlin/realLife/SingletonKotlinRealLife/src/main/kotlin/EventService.kt
typealias Listener = (value: String) -> Unit;

data class Unsubscribe(val unsubscribe: () -> Unit)

class EventService private constructor() {
  companion object {
    @Volatile
    private lateinit var instance: EventService;

    fun getInstance(): EventService {
      synchronized(this) {
        if (!::instance.isInitialized) {
          instance = EventService();
        }
        return instance;
      }
    }
  }
  
  private var listeners: MutableList<Listener> = arrayListOf()
  
  fun emit(value: String) {
    this.notifyListeners(value)
  }

  fun subscribe(listener: Listener): Unsubscribe {
    listeners.add(listener)
    return Unsubscribe { listeners.remove(listener) }
  }
  
  private fun notifyListeners(value: String) {
    listeners.forEach { listener -> listener(value) }
  } 
}

File: #3.design-patterns/creational/singleton/typescript/src/realLife/EventService.ts
type Listener = (value: string) => void;

type Unsubscribe = { unsubscribe: () => void };

export class EventService {
  private listeners: Listener[] = [];

  protected static _instance: EventService | null = null;

  private constructor() {}

  static get instance(): EventService {
    if (EventService._instance === null) {
      EventService._instance = new EventService();
    }

    return EventService._instance;
  }

  public emit(value: string) {
    this.notifyListeners(value);
  }

  public subscribe(listener: Listener): Unsubscribe {
    this.listeners.push(listener);
    return {
      unsubscribe: () => {
        this.listeners = this.listeners.filter(
          (anotherListener) => anotherListener !== listener
        );
      },
    };
  }

  private notifyListeners(value: string) {
    this.listeners.forEach((listener) => listener(value));
  }
}
